### ISSUE TITLE
[v2] mime sniffing local resources can result in browser evaluating them in wrong context

### ISSUE THREAD

**Issue by napalu** [wailsapp/wails#1300]
**Is your feature request related to a problem? Please describe.**
Just was confronted with the following scenario: clone pdf.js repository and build a local dist to serve the repo locally in wails following the instructions at https://github.com/mozilla/pdf.js to embed a local pdf viewer. The problem: 
- some HTML files like viewer.html have a leading BOM (utf-8 byte order mark) leading them to be characterized as `text/plain` instead of `text/html` with the result that the browser may display them as plain text
- some SVG files start with a comment such as `<!--- comment -->` leading them to be characterized as `text/html` instead of `image/svg+xml` with the result that the svg files may not be displayed properly.
  

**Describe the solution you'd like**
For html files we could check if BOM is present at the start of file and if so pass the slice after the BOM to the mime detect functions. For SVG files we could check for a comment section at the start of file and if present pass the slice after the end of the comment section. 

Alternatively, we could short-circuit detection of .html and .svg files in the same way that `mimecache.go` currently does in `mimecache.GetMimeType` for .js and .css files. 

For security and future compatibility reasons I'd prefer testing for the BOM in .html and for comments in .svg files and  allow `mimetype.detect` and `http.DetectContentType` to sniff the the mime-type.  

**Describe alternatives you've considered**
Manually remove BOM and eventual comments from resources is possible but I can think of situations where one could register additional (user-defined) custom uri schemes which would access resources other than embedded ones (for example from the file system at large) for display in the browser.

**Additional context**
Example where viewer.html starts with a BOM is embedded in an iframe :

![image](https://user-images.githubusercontent.com/6690378/160855944-02d9ce19-b54b-43a5-9676-f39e3366ad17.png)

Example where viewer.html BOM is removed/short-circuited but some SVG files with leading comments are passed as `text/html` instead of `image/svg+xml`:

![image](https://user-images.githubusercontent.com/6690378/160856652-eadb6aae-4e60-4b17-b5dc-b9dd718e91d1.png)

Example of what the viewer looks like with all .html and .svg files evaluated in the correct context:

![image](https://user-images.githubusercontent.com/6690378/160857254-4b956572-1473-4253-8a28-9719edcd0230.png)

**Comment by leaanthony** [comment]
Ah brilliant! Mime type detection is way harder than I first thought. Even using 2 dedicated methods appears to not work 😀 Thanks for working this out. Will review PR

**Comment by stffabi** [comment]
The SVG comment problem might already be fixed upstream https://github.com/gabriel-vasile/mimetype/pull/193 Maybe we could bump the dependency and see if it fixes at least that problem.

**Comment by napalu** [comment]
@leaanthony I'm testing the upstream - the fixes are in master but are not yet in the current 1.40 release so we'd have to wait until a new release is done. I don't mind re-purposing the PR but I'd like to make sure we can get the functionality. I'll post here with the results.

**Comment by napalu** [comment]
Here's the result of the tests. With the current `github.com/gabriel-vasile/mimetype v1.3.1`:
```
--- FAIL: TestGetMimetype (0.00s)
    --- FAIL: TestGetMimetype/html-bom-utf8 (0.00s)
        mimecache_test.go:42: GetMimetype() = 'text/plain; charset=utf-8', want 'text/html; charset=utf-8'
    --- FAIL: TestGetMimetype/svg-w-comment (0.00s)
        mimecache_test.go:42: GetMimetype() = 'text/html; charset=utf-8', want 'image/svg+xml'
    --- FAIL: TestGetMimetype/svg-w-control-comment (0.00s)
        mimecache_test.go:42: GetMimetype() = 'text/html; charset=utf-8', want 'image/svg+xml'
FAIL
```
 with the latest current release v1.4.0:

```
--- FAIL: TestGetMimetype (0.00s)
    --- FAIL: TestGetMimetype/html-bom-utf8 (0.00s)
        mimecache_test.go:42: GetMimetype() = 'text/plain; charset=utf-8', want 'text/html; charset=utf-8'
    --- FAIL: TestGetMimetype/svg-w-comment (0.00s)
        mimecache_test.go:42: GetMimetype() = 'text/html; charset=utf-8', want 'image/svg+xml'
    --- FAIL: TestGetMimetype/svg-w-control-comment (0.00s)
        mimecache_test.go:42: GetMimetype() = 'text/html; charset=utf-8', want 'image/svg+xml'
FAIL
```

interestingly both versions pass the `svgWithBomCommentAndControlChars` test

the latest upstream (not yet released) only fails on html BOM:

```
--- FAIL: TestGetMimetype (0.00s)
    --- FAIL: TestGetMimetype/html-bom-utf8 (0.00s)
        mimecache_test.go:42: GetMimetype() = 'text/plain; charset=utf-8', want 'text/html; charset=utf-8'
FAIL
```
Proposal: @stffabi  and @leaanthony  - how about we merge the current PR and check/revert when the latest upstream is released?

**Comment by napalu** [comment]
Here's the result of the tests. With the current `github.com/gabriel-vasile/mimetype v1.3.1`:
```
--- FAIL: TestGetMimetype (0.00s)
    --- FAIL: TestGetMimetype/html-bom-utf8 (0.00s)
        mimecache_test.go:42: GetMimetype() = 'text/plain; charset=utf-8', want 'text/html; charset=utf-8'
    --- FAIL: TestGetMimetype/svg-w-comment (0.00s)
        mimecache_test.go:42: GetMimetype() = 'text/html; charset=utf-8', want 'image/svg+xml'
    --- FAIL: TestGetMimetype/svg-w-control-comment (0.00s)
        mimecache_test.go:42: GetMimetype() = 'text/html; charset=utf-8', want 'image/svg+xml'
FAIL
```
 with the latest current release v1.4.0:

```
--- FAIL: TestGetMimetype (0.00s)
    --- FAIL: TestGetMimetype/html-bom-utf8 (0.00s)
        mimecache_test.go:42: GetMimetype() = 'text/plain; charset=utf-8', want 'text/html; charset=utf-8'
    --- FAIL: TestGetMimetype/svg-w-comment (0.00s)
        mimecache_test.go:42: GetMimetype() = 'text/html; charset=utf-8', want 'image/svg+xml'
    --- FAIL: TestGetMimetype/svg-w-control-comment (0.00s)
        mimecache_test.go:42: GetMimetype() = 'text/html; charset=utf-8', want 'image/svg+xml'
FAIL
```

interestingly both versions pass the `svgWithBomCommentAndControlChars` test

the latest upstream (not yet released) only fails on html BOM:

```
--- FAIL: TestGetMimetype (0.00s)
    --- FAIL: TestGetMimetype/html-bom-utf8 (0.00s)
        mimecache_test.go:42: GetMimetype() = 'text/plain; charset=utf-8', want 'text/html; charset=utf-8'
FAIL
```
Proposal: @stffabi  and @leaanthony  - how about we merge the current PR and check/revert when the latest upstream is released?

**Comment by leaanthony** [comment]
I've actually forked this module and updated master to use it. If we fix the issues there, we can push the changes upstream but not have to wait for them. Thoughts?

**Comment by napalu** [comment]
You mean integrate latest `github.com/gabriel-vasile/mimetype` head in wails - fix the html BOM issue, open an issue upstream to  github.com/gabriel-vasile/mimetype` and push an associated PR? Works for me as well - all I want is the functionality :) We should keep the tests though.

**Comment by napalu** [comment]
You mean integrate latest `github.com/gabriel-vasile/mimetype` head in wails - fix the html BOM issue, open an issue upstream to  github.com/gabriel-vasile/mimetype` and push an associated PR? Works for me as well - all I want is the functionality :) We should keep the tests though.

**Comment by leaanthony** [comment]
Realised I'd not pushed the changes, so done so now. Feel free to open a PR on https://github.com/wailsapp/mimetype for the BOM issue and we can get this fixed 👍

**Comment by napalu** [comment]
@leaanthony I just pushed the pull request to your fork - feel free to merge :)

**Comment by leaanthony** [comment]
I've merged that PR. Feel free to update go.mod in Wails to pull it in. You'll need to update using `go get -u github.com/wailsapp/mimetype@latestcommithash`. I'm afk currently

**Comment by gabriel-vasile** [comment]
The [correct fix](https://github.com/gabriel-vasile/mimetype/pull/268) for this issue is released in [mimetype v1.4.1-rc1](https://github.com/gabriel-vasile/mimetype/releases/tag/v1.4.1-rc1). Thank you again, @napalu.
